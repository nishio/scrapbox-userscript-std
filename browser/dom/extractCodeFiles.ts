import type { Line } from "@cosense/types/userscript";

/** Represents a single source code file with its code blocks */
export interface CodeFile {
  /** file name */
  filename: string;

  /** language */
  lang: string;

  /** splitted code */
  blocks: CodeBlock[];
}

/** Represents a single code block within a source file */
export interface CodeBlock {
  /** ID of the first line in the code block */
  startId: string;

  /** ID of the last line in the code block */
  endId: string;

  /** Last update timestamp of the code block */
  updated: number;

  /** Indentation level of the .code-title element in Scrapbox */
  indent: number;

  /** Lines of code within the block
   *
   * Excludes `.code-title`
   *
   * Indentation is already removed from each line
   */
  lines: string[];
}

/** `scrapbox.Page.lines`からcode blocksを取り出す
 *
 * @param lines ページの行
 * @return filenameをkeyにしたソースコードのMap
 */
export const extractCodeFiles = (
  lines: Iterable<Line>,
): Map<string, CodeFile> => {
  const files = new Map<string, CodeFile>();

  for (const line of lines) {
    if (!("codeBlock" in line)) continue;
    const { filename, lang, ...rest } = line.codeBlock;
    const file = files.get(filename) ?? { filename, lang, blocks: [] };
    if (rest.start || file.blocks.length === 0) {
      file.blocks.push({
        startId: line.id,
        endId: line.id,
        updated: line.updated,
        // Register the indentation level of `.code-title`, not the content
        indent: rest.indent - 1,
        lines: [],
      });
    } else {
      const block = file.blocks[file.blocks.length - 1];
      block.endId = line.id;
      block.updated = Math.max(block.updated, line.updated);
      block.lines.push([...line.text].slice(block.indent + 1).join(""));
    }

    files.set(filename, file);
  }

  return files;
};
