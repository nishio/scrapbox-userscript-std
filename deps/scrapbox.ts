export type {
  BaseLine,
  Line,
  Scrapbox,
} from "https://raw.githubusercontent.com/scrapbox-jp/types/0.5.0/userscript.ts";
export type {
  BaseStore,
} from "https://raw.githubusercontent.com/scrapbox-jp/types/0.5.0/baseStore.ts";
export * from "https://esm.sh/@progfay/scrapbox-parser@9.0.0";
