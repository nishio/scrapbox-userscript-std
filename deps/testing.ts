export * from "https://deno.land/std@0.217.0/testing/asserts.ts";
export * from "https://deno.land/std@0.217.0/testing/snapshot.ts";
