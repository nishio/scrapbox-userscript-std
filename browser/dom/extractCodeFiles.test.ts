import { extractCodeFiles } from "./extractCodeFiles.ts";
import { Line } from "../../deps/scrapbox.ts";
import { assertSnapshot } from "../../deps/testing.ts";
import sample from "./sample-lines1.json" assert { type: "json" };

Deno.test("extractCodeFiles", async (t) => {
  await assertSnapshot(
    t,
    extractCodeFiles(sample as Line[]),
  );
});
