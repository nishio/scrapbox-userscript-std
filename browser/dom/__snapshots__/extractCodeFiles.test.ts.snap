export const snapshot = {};

snapshot[`extractCodeFiles 1`] = `
Map(5) {
  "main.cpp" => {
    blocks: [
      {
        endId: "63b7b1261280f00000c9bc36",
        indent: 2,
        lines: [
          "#include <iostream>",
          "",
        ],
        startId: "63b7b1261280f00000c9bc34",
        updated: 1672982822,
      },
      {
        endId: "63b7b1261280f00000c9bc3c",
        indent: 2,
        lines: [
          "int main() {",
          '  std::cout << "Hello, C++" << "from scrapbox.io" << std::endl;',
          "}",
          "",
        ],
        startId: "63b7b1261280f00000c9bc38",
        updated: 1672982822,
      },
    ],
    filename: "main.cpp",
    lang: "cpp",
  },
  "py" => {
    blocks: [
      {
        endId: "63b7b1261280f00000c9bc2a",
        indent: 0,
        lines: [
          'print("Hello World!")',
        ],
        startId: "63b7b1261280f00000c9bc29",
        updated: 1672982822,
      },
    ],
    filename: "py",
    lang: "py",
  },
  "python" => {
    blocks: [
      {
        endId: "63b7b1261280f00000c9bc31",
        indent: 1,
        lines: [
          \`console.log("I'm JavaScript");\`,
        ],
        startId: "63b7b1261280f00000c9bc30",
        updated: 1672982822,
      },
    ],
    filename: "python",
    lang: "js",
  },
  "インデント.md" => {
    blocks: [
      {
        endId: "63b7b1261280f00000c9bc2e",
        indent: 1,
        lines: [
          "- インデント",
          "  - インデント",
        ],
        startId: "63b7b1261280f00000c9bc2c",
        updated: 1672982822,
      },
    ],
    filename: "インデント.md",
    lang: "md",
  },
  "コードブロック.py" => {
    blocks: [
      {
        endId: "63b7b1261280f00000c9bc27",
        indent: 0,
        lines: [
          'print("Hello World!")',
        ],
        startId: "63b7b1261280f00000c9bc26",
        updated: 1672982822,
      },
    ],
    filename: "コードブロック.py",
    lang: "py",
  },
}
`;
